### PR TITLE
Fix Spark 3.4 build issues

### PR DESCRIPTION
**Pull request by nartal1**
This closes https://github.com/NVIDIA/spark-rapids/issues/6671, closes https://github.com/NVIDIA/spark-rapids/issues/5806 and contributes to https://github.com/NVIDIA/spark-rapids/issues/6851

This PR fixes few build issues : 

1. https://github.com/NVIDIA/spark-rapids/issues/6671 -> Change messageParmeters from Array to Map for Spark-3.4
2. https://github.com/NVIDIA/spark-rapids/issues/5806 -> Added argument to QueryExecutionErrors 
3. https://github.com/NVIDIA/spark-rapids/issues/6851 -> Fixed RapidsUtilErrors compile errors. Created a shim to include new type i.e it was changed from String to SQLQueryContext.

Before:
```
[ERROR] [Error] /home/nartal/workspace/spark-rapids/sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala:28: not enough arguments for method invalidElementAtIndexError: (index: Int, numElements: Int, context: org.apache.spark.sql.catalyst.trees.SQLQueryContext)ArrayIndexOutOfBoundsException.
Unspecified value parameter context.
[ERROR] [Error] /home/nartal/workspace/spark-rapids/sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala:30: not enough arguments for method invalidArrayIndexError: (index: Int, numElements: Int, context: org.apache.spark.sql.catalyst.trees.SQLQueryContext)ArrayIndexOutOfBoundsException.
Unspecified value parameter context.
[ERROR] [Error] /home/nartal/workspace/spark-rapids/sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala:57: type mismatch;
 found   : String
 required: org.apache.spark.sql.catalyst.trees.SQLQueryContext
[ERROR] [Error] /home/nartal/workspace/spark-rapids/sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala:65: type mismatch;
 found   : String
 required: org.apache.spark.sql.catalyst.trees.SQLQueryContext
[ERROR] [Error] /home/nartal/workspace/spark-rapids/sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala:71: type mismatch;
 found   : String
 required: org.apache.spark.sql.catalyst.trees.SQLQueryContext
[ERROR] [Error] /home/nartal/workspace/spark-rapids/sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/SparkUpgradeExceptionShims.scala:29: type mismatch;
 found   : Array[String]
 required: Map[String,String]
[INFO] [Info] : Array[String] <: Map[String,String]?

[ERROR] 35 errors found

```

After:
```
[ERROR] 29 errors found
```

### CODE REVIEW

**Comment by nartal1** [comment]
build

**Comment by nartal1** [comment]
build

**Comment by nartal1** [comment]
build

File: sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtilsFor33Xand34X.scala [hunk 1]
```diff
@@ -19,17 +19,9 @@ package org.apache.spark.sql.rapids.shims
...
-object RapidsErrorUtils {
-  def invalidArrayIndexError(index: Int, numElements: Int,
-      isElementAtF: Boolean = false): ArrayIndexOutOfBoundsException = {
-    if (isElementAtF) {
-      QueryExecutionErrors.invalidElementAtIndexError(index, numElements)
-    } else {
-      QueryExecutionErrors.invalidArrayIndexError(index, numElements)
-    }
-  }
+trait RapidsErrorUtilsFor33Xand34X {
```
Comment (HaoYang670) [inline sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtilsFor33Xand34X.scala:24]: nit: Do you think `RapidsErrorUtilsFor330plus` is a better name?

**Review by HaoYang670 — DISMISSED** [review]
Using trait is a good implementation 👍 here to avoid copying the code.
Maybe we could also introduce this pattern in the shims from 311 to 32X

File: pom.xml [hunk 2]
```diff
@@ -898,6 +898,7 @@
                                 <configuration>
                                     <sources>
                                         <source>${project.basedir}/src/main/340/scala</source>
```
[the remark this answers is not in the document]
Comment (nartal1) [reply pom.xml:900]: I agree we could move all files which are under `340 ` directory to `340+` .  I just tested it. Is it okay if I include this change in this PR?

[the remark this answers is not in the document]
Comment (jlowe) [reply pom.xml:900]: I think we should move the files that apply to >= 3.4.0 as part of this, but there are a few files that need to remain in a 340 directory as they are specific to 3.4.0, e.g.: the 3.4.0 version of SparkShimServiceProvider

**Review by tgravescs — APPROVED** [review]

**Comment by tgravescs** [comment]
build

File: pom.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (nartal1) [reply pom.xml:900]: Thanks! I referred to 330 shim and kept the files that are specific to 340/ directory and moved the remaining files to 340+/ directory. PTAL.

File: sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtilsFor33Xand34X.scala (hunk 1 above)
Comment (nartal1) [reply sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtilsFor33Xand34X.scala]: Done

**Review by jlowe — APPROVED** [review]

**Pull request merged**